test/Autofac.Test/Core/ComponentRegistrationTests.cs

```diff
@@ -86,4 +86,71 @@ public void AttachingToPipelineBuildingShouldFailIfAlreadyBuilt()
 
         Assert.Throws<InvalidOperationException>(() => registration.PipelineBuilding += (s, p) => { });
     }
+
```

Comment:
Would these be part of `Autofac.Specification.Test` since it'd be supported public API?

Reply:
I don't have a super-strong feeling about it; is that the rule, that any tests for publicly visible API endpoints should be in `Specification.Test`? I pretty much put those tests in here because they were compact, could use the `Factory` class and tested _only_ `ComponentRegistration`, without involving the `Registered` event.

I'm going to add a new test file to `Specification.Test` to cover off `IStartable`, decorators, etc. Perhaps that's enough for defining the public surface?

Reply:
Generally what I was trying to do with the `Specification.Test` is separate out "tests _for_ the public API tested _by_ the public API like an end user would consume them" from the other kinds of tests ("tests for X that use internals" or "tests for internal APIs") so we would know during a refactor which tests _should_ be sort of safe to update vs. which ones really shouldn't be touched and should continue to work. Also, from an end user perspective, it's easier to search and find "real examples" if you don't have to determine whether we were using something internal to show the workings of things.

Tests like "here's how to consume this thing from a module" that are almost documentation seemed like they fit under the public API spec test sort of thing.

We never did actually get all of the tests separated out. It was too big of a job to do all at once, so it was more iterative - "if you're touching this thing or adding new tests, put them in the right spot so we don't have to figure it out later."

Reply:
Noted, ok, will move some things around.